Extensions: Add methods to override to public headers

flovouin opened the pull request:
Hi,

I've been using YapDatabase in Swift in production for a while now, and so far it has filled all my needs. When I was looking for synchronisation options, `CloudCore` was an obvious choice and has really made my life easier. However to make it work in Swift I needed to add methods to override to public headers, as otherwise I simply couldn't override them.

This PR does exactly this: no code changes, just declarations. This branch is what I've been using in production for months now, and so far it's worked great for me. It'd be nice if this were part of the official code base, instead of me relying on my fork and having to update it once in a while.

I've taken care of only exposing methods with documentation specifying a required override, but you might want to double check these, and whether it might be beneficial to turn other declarations public.

Edit: Here's a [small project](https://github.com/flovouin/CloudCoreSwift) showing what it looks like in Swift. If you change the `Podfile` to use the official code base, the project will not compile.

Cheers,

robbiehanson commented:
You are correct in pointing out that this is a problem for Swift. Thank you, I wasn't aware of this problem. However, I don't think this is the best solution we can come up with.

If you take a look at the compiled framework for YapDatabase, you'll notice there's a folder called "PrivateHeaders". Inside this folder is the file you're looking for: YapDatabaseCloudCorePrivate.h

If your code can import this file, then all your problems go away. And indeed, there's zero problems accomplishing this task in Objective-C. Nothing fancy to do, just add this line:

```
#import <YapDatabase/YapDatabaseCloudCorePrivate.h>
```

Done. But if we try to do something similar in Swift it doesn't work.

So, to summarize my position:
- I don't think the correct thing to do is to expose private/internal/advanced methods for all users
- Instead I think this should be opt-in
- And the opt-in should simply be an additional import that has the word "Private" somewhere

So we just need to solve the "PrivateHeaders" issue for Swift...

It seems that other people have run into similar problems: https://github.com/CocoaPods/CocoaPods/issues/4656

And, from what I understand from the above link, the answer may be to switch to a module.modulemap file. I.e. instead of allowing CocoaPods to generate the modulemap for us, we manually create it ourself. This will allow us to add "explicit" (opt-in) private submodules.

In fact, there are already modulemaps in the repo:
https://github.com/yapstudios/YapDatabase/blob/master/Framework/Mac/module.modulemap

So we would just need to:

- Change the podspec to use the modulemap
- Add Private sections for each extension

flovouin commented:
Hi @robbiehanson,

Thank you for your reply!

Yes, I understand your concern and agree that there should be a better solution.

Modifying the modules makes sense, and seems to be the solution adopted both by other Pods (as you linked) and Apple, which, [according to Stackoverflow](https://stackoverflow.com/questions/3725857/protected-methods-in-objective-c), seems to be doing [something similar](https://developer.apple.com/documentation/uikit/touches_presses_and_gestures/implementing_a_custom_gesture_recognizer?language=objc).

Please look at the updated PR, which should be closer to what you're expecting. However, I'm still not entirely confident about my changes, so feel free to comment, or to close this and make the changes yourself if you'd prefer!

A few things:

- I've only made the changes to the `CloudCore` extension. Indeed, as far as I know, it is the only extension for which subclassing is actually recommended in the docs. I've used views and full text search for a while now, and never felt the need to subclass them.
- Obviously, I also had to expose the extensions protocol internals through another submodule. One particular import (`YapMemoryTable.h`) meant that I'd also have to expose core internals, which didn't seem like a good idea. It looked like the import wasn't used, so I removed it from `YapDatabaseExtensionPrivate.h` to solve the problem.
- Finally, I had to remove the relevant private headers (exposed in submodules) from the `Podspec` `private_header_files`. This is one of the things I'm less confident about.

All in all, my tests seems to be conclusive: I can successfully override the relevant methods in Swift, but only if I import the newly defined submodules: `YapDatabaseExtensionPrivate` and `YapDatabaseCloudCorePrivate`. I've also updated my example project to reflect these changes.

Cheers,

robbiehanson commented:
> Apple [...] seems to be doing something similar.

Nice find. It looks like they've gone as far as splitting things into 3: public, subclass & private. Hmm... I think we're fine just exposing the private stuff for now. If we want to get extra fancy later, we can go that extra mile.

> as far as I know, it is the only extension for which subclassing is actually recommended in the docs.

That's true. And I've never had a need to subclass anything else either.

> Finally, I had to remove the relevant private headers [...] This is one of the things I'm less confident about.

I looked into this, and I think you did the right thing here. It's true that this puts some "xPrivate.h" files into the CocoaPods Headers/Public folder, but that seems to be our only option here (in order to expose them to Swift).

But I don't think it's a problem. The only way for someone to access those file is if they explicitly `import` them. So no problems there.

> I can successfully override the relevant methods in Swift, but only if I import the newly defined submodules: `YapDatabaseExtensionPrivate` and `YapDatabaseCloudCorePrivate`

Nice! I took this one step further for you:

```
explicit module YapDatabaseCloudCorePrivate {
    ...
  	
    export YapDatabaseExtensionPrivate
}
```

So now you only need to import `YapDatabaseCloudCorePrivate`.

flovouin commented:
Thanks a lot for the merge and the feedback!

robbiehanson commented:
@flovouin CocoaPods is giving us grief over the modulemap. Try running this command:
```
pod spec lint --subspec=YapDatabase/Standard/Extensions/Relationships --fail-fast --verbose --no-clean
```
It fails with the following:
```
module.modulemap:33:10: error: header 'YapDatabaseView.h' not found
```

Here's what's happening:
- YapDatabase.podspec is configured to allow people to pick-and-choose which extensions they want
- However, our podspec is now using a fixed modulemap which includes ALL extensions
- The compiler is unable to load the YapDatabase modulemap, and so generates an error

I think I might take a stab at splitting YapDatabaseCloudCore into private vs subclass stuff...

robbiehanson commented:
OK, here's the compromise for now (until I separate the private from subclass stuff). There are now 2 new targets with the Podspec:

- Standard+ModuleMap
- SQLCipher+ModuleMap

If you use these, it will use the modulemap (as was intended previously). So something like this should work:

`pod 'YapDatabase/SQLCipher+ModuleMap'`
